target/target.go

```diff
@@ -47,47 +46,130 @@ var (
 	return &Target{Config: cfg, Body: body, Url: url, Method: method}
 }
 
-// Get constructs the list of work to be completed
-func (t *Target) GetWork() *requester.Work {
-	// request
-	req, err := http.NewRequest(t.Method, t.Url, nil)
+func NewTargetFromOptions(baseUrl string, opts ...OptionFunc) (*Target, error) {
+	copyCfg := defaultCfg
+	cfg := &copyCfg
+	var err error
+	for _, opt := range opts {
+		err = with(cfg, opt, err)
+	}
+	body := buildBody(cfg.BodyConfig)
+	url := strings.TrimSuffix(baseUrl, "/") + cfg.Endpoint
+	return &Target{Config: cfg, Body: body, Url: url, Method: "POST"}, err
+}
+
+type OptionFunc func(*Config) error
+
+func with(c *Config, f OptionFunc, err error) error {
 	if err != nil {
-		panic(err)
+		return err
 	}
+	return f(c)
+}
 
-	// headers
-	for header, values := range t.Config.Header {
-		for _, v := range values {
-			req.Header.Add(header, v)
-		}
+func RunTimeout(s string) OptionFunc {
+	return func(c *Config) error {
+		var err error
+		c.RunTimeout, err = time.ParseDuration(s)
+		return err
+	}
+}
+
+func RequestTimeout(s string) OptionFunc {
+	return func(c *Config) error {
+		timeout, err := time.ParseDuration(s)
+		c.RequestTimeout = int(timeout.Nanoseconds())
+		return err
+	}
+}
+
+func NumAgents(s string) OptionFunc {
+	return func(c *Config) error {
+		var err error
+		c.NumAgents, err = strconv.Atoi(s)
+		return err
+	}
+}
+
+func Throttle(s string) OptionFunc {
+	return func(c *Config) error {
+		throttle, err := strconv.Atoi(s)
+		c.Throttle = float64(throttle)
+		return err
+	}
+}
+
+func Stream(s string) OptionFunc {
+	return func(c *Config) error {
+		c.Stream = s == ""
+		return nil
+	}
+}
+
+func NumErrors(s string) OptionFunc {
+	return func(c *Config) error {
+		var err error
+		c.NumErrors, err = strconv.Atoi(s)
+		return err
+	}
+}
+
+func NumTransactions(s string) OptionFunc {
+	return func(c *Config) error {
+		var err error
+		c.NumTransactions, err = strconv.Atoi(s)
+		return err
+	}
+}
+
+func NumSpans(s string) OptionFunc {
+	return func(c *Config) error {
+		var err error
+		c.NumSpans, err = strconv.Atoi(s)
+		return err
+	}
+}
+
+func NumFrames(s string) OptionFunc {
+	return func(c *Config) error {
+		var err error
+		c.NumFrames, err = strconv.Atoi(s)
+		return err
 	}
+}
+
+// Returns a runnable that simulates APM agents sending requests to APM Server with the `target` configuration
+// Mutates t.Body (for compression) and t.Headers
+func (t *Target) GetWork() *requester.Work {
+
 	// Use the defaultUserAgent unless the Header contains one, which may be blank to not send the header.
-	if _, ok := req.Header["User-Agent"]; !ok {
-		req.Header.Add("User-Agent", defaultUserAgent)
+	if _, ok := t.Config.Header["User-Agent"]; !ok {
+		t.Config.Header.Add("User-Agent", defaultUserAgent)
 	}
 
 	if len(t.Body) > 0 {
-		req.Header.Add("Content-Type", "application/x-ndjson")
+		t.Config.Header.Add("Content-Type", "application/x-ndjson")
```

Comment:
Using `t.Config.Stream` here would be more accurate.

Reply:
i don't get you, using `t.Config.Stream` instead of what?

Reply:
Ah that was actually a comment I thought I had deleted, please disregard.